awswrangler/athena/_spark.py

```diff
@@ -0,0 +1,210 @@
+
+    Returns
+    -------
+        Session id
+    """
+    client_athena = _utils.client(service_name="athena", session=boto3_session)
+    engine_configuration: "EngineConfigurationTypeDef" = {
+        "CoordinatorDpuSize": coordinator_dpu_size,
+        "MaxConcurrentDpus": max_concurrent_dpus,
+        "DefaultExecutorDpuSize": default_executor_dpu_size,
+    }
+    if additional_configs:
+        engine_configuration["AdditionalConfigs"] = additional_configs
+    response = client_athena.start_session(
+        WorkGroup=workgroup,
+        EngineConfiguration=engine_configuration,
+        # NotebookId=notebook_id,
+        # NotebookVersion=notebook_version if notebook_id else None,
+        SessionIdleTimeoutInMinutes=idle_timeout,
+    )
+    _logger.info("Session info:\n%s", response)
+    session_id: str = response["SessionId"]
+    # Wait for the session to reach IDLE state to be able to accept calculations
+    _wait_session(
+        session_id=session_id,
+        boto3_session=boto3_session,
+    )
+    return session_id
+
+
+def run_spark_calculation(
+    code: str,
+    workgroup: str,
+    session_id: Optional[str] = None,
+    notebook_id: Optional[str] = None,
+    notebook_version: str = "Athena notebook version 1",
+    coordinator_dpu_size: int = 1,
+    max_concurrent_dpus: int = 5,
+    default_executor_dpu_size: int = 1,
+    additional_configs: Optional[Dict[str, Any]] = None,
+    idle_timeout: int = 15,
+    boto3_session: Optional[boto3.Session] = None,
+) -> "GetCalculationExecutionResponseTypeDef":
+    """
+    Execute Spark Calculation and wait for completion.
+
+    Parameters
+    ----------
+    code : str
+    workgroup : str
+    session_id : str, optional
+    notebook_id : str, optional
+    notebook_version : str, optional
+    coordinator_dpu_size : int, optional
+    max_concurrent_dpus : int, optional
+    default_executor_dpu_size: int, optional
+    additional_configs : Dict[str, Any], optional
+    idle_timeout : int, optional
+    boto3_session : boto3.Session(), optional
+
+    Returns
+    -------
+        Calculation response
+    """
+    client_athena = _utils.client(service_name="athena", session=boto3_session)
+
+    session_id = (
+        create_spark_session(
+            workgroup=workgroup,
+            notebook_id=notebook_id,
+            notebook_version=notebook_version,
+            coordinator_dpu_size=coordinator_dpu_size,
+            max_concurrent_dpus=max_concurrent_dpus,
+            default_executor_dpu_size=default_executor_dpu_size,
+            additional_configs=additional_configs,
+            idle_timeout=idle_timeout,
+            boto3_session=boto3_session,
+        )
+        if not session_id
+        else session_id
+    )
+
+    response = client_athena.start_calculation_execution(
+        SessionId=session_id,
+        CodeBlock=code,
+    )
+    _logger.info("Calculation execution info:\n%s", response)
+
+    return _get_calculation_execution_results(
```

Comment:
Can the return value of an Athena Spark execution be a DataFrame? Or will the output always just be written to an S3 location?

I'm mainly just wondering if there's a way to sensibly make these API calls accept or return Data Frames.

Reply:
I am looking into the ways to support that where it is applicable but that is very much dependent on what spark code you are running. 

There is json metadata file in the results path along with stdout/err text files, but it's almost always empty, well at least have been in my tests so far. Documentation is a bit lacking on this. I'll play around to see what we can make of it.